### PR TITLE
feat: make metrics build specific

### PR DESCRIPTION
**Pull request by thugrock7**
## Description
CPU and memory metrics which are added, making them linux specific. For other platforms using no-op

### Testing
Tested on Mac, linux

### CODE REVIEW

File: instrumentation/opentelemetry/internal/metrics/metricsImpl.go
Comment (varkey98) [inline instrumentation/opentelemetry/internal/metrics/metricsImpl.go:1]: rename to linux_metrics.go

File: instrumentation/opentelemetry/internal/metrics/metricsNoop.go
Comment (varkey98) [inline instrumentation/opentelemetry/internal/metrics/metricsNoop.go:1]: noop_metrics.go

File: instrumentation/opentelemetry/internal/metrics/metrics.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+	}
+	memory, err := meter.Float64ObservableGauge("hypertrace.agent.memory", metric.WithDescription("Metric to monitor memory usage"))
+	if err != nil {
+		return fmt.Errorf("error while setting up memory metric counter: %v", err)
+	}
+	// Register the callback function for both cpu_seconds and memory observable gauges
+	_, err = meter.RegisterCallback(
+		func(ctx context.Context, result metric.Observer) error {
+			sysMetrics := newSystemMetrics()
+			err := sysMetrics.getCurrentMetrics()
```
Comment (varkey98) [inline instrumentation/opentelemetry/internal/metrics/metrics.go:44]: now that I see these 2 together, lets merge the logic of getCurrentMetrics to the constructor and remove that function from the interface itself

**Comment by varkey98** [comment]
done reviewing, missed this earlier

File: instrumentation/opentelemetry/internal/metrics/metrics.go [hunk 2]
```diff
@@ -0,0 +1,57 @@
+package metrics
+
+import (
+	"context"
+	"fmt"
+	"go.opentelemetry.io/otel"
```
Comment (varkey98) [inline instrumentation/opentelemetry/internal/metrics/metrics.go:6]: pls sort imports here

**Comment by varkey98** [comment]
@shashank11p @puneet-traceable pls take a look as well

**Comment by varkey98** [comment]
Adding some context here: The reason is that `/proc/pid/stat` file is only available on linux and trying to read from that file on other platforms results in an error

File: instrumentation/opentelemetry/internal/metrics/linux_metrics.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+func (lm *linuxMetrics) getCPU() float64 {
+	return lm.cpuSecondsTotal
+}
+
+func (lm *linuxMetrics) processStatsFromPid(pid int) (*processStats, error) {
+	procFilepath := filepath.Join("/proc", strconv.Itoa(pid), "stat")
+	var err error
+	if procStatFileBytes, err := os.ReadFile(filepath.Clean(procFilepath)); err == nil {
+		if stat, err := lm.parseProcStatFile(procStatFileBytes, procFilepath); err == nil {
+			if err != nil {
```
Comment (puneet-traceable) [inline instrumentation/opentelemetry/internal/metrics/linux_metrics.go:59]: what is this error check? above check is `err == nil`

File: instrumentation/opentelemetry/internal/metrics/linux_metrics.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+func (lm *linuxMetrics) getCPU() float64 {
+	return lm.cpuSecondsTotal
+}
+
+func (lm *linuxMetrics) processStatsFromPid(pid int) (*processStats, error) {
+	procFilepath := filepath.Join("/proc", strconv.Itoa(pid), "stat")
+	var err error
+	if procStatFileBytes, err := os.ReadFile(filepath.Clean(procFilepath)); err == nil {
```
Comment (puneet-traceable) [inline instrumentation/opentelemetry/internal/metrics/linux_metrics.go:57]: better to return back early if `err != nil`

File: instrumentation/opentelemetry/internal/metrics/linux_metrics.go [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+	}
+	stat, err := lm.parseProcStatFile(procStatFileBytes, procFilepath)
+	if err != nil {
+		return nil, err
+	}
+	return stat, nil
+}
+
+// ref: /proc/pid/stat section of https://man7.org/linux/man-pages/man5/proc.5.html
+func (lm *linuxMetrics) parseProcStatFile(bytesArr []byte, procFilepath string) (*processStats, error) {
```
Comment (varkey98) [inline instrumentation/opentelemetry/internal/metrics/linux_metrics.go:69]: can you add some unit tests for this file as well?

File: instrumentation/opentelemetry/internal/metrics/linux_metrics.go [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+	return lm.memory
+}
+
+func (lm *linuxMetrics) getCPU() float64 {
+	return lm.cpuSecondsTotal
+}
+
+func (lm *linuxMetrics) processStatsFromPid(pid int) (*processStats, error) {
+	procFilepath := filepath.Join("/proc", strconv.Itoa(pid), "stat")
+	var err error
```
Comment (varkey98) [inline instrumentation/opentelemetry/internal/metrics/linux_metrics.go:56]: this also can be removed.

**Review by varkey98 — APPROVED** [review]

**Comment by varkey98** [comment]
@thugrock7 Pls test it on windows as well

File: instrumentation/opentelemetry/internal/metrics/linux_metrics.go [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+
+func (lm *linuxMetrics) getMemory() float64 {
+	return lm.memory
+}
+
+func (lm *linuxMetrics) getCPU() float64 {
+	return lm.cpuSecondsTotal
+}
+
+func (lm *linuxMetrics) processStatsFromPid(pid int) (*processStats, error) {
```
Comment (puneet-traceable) [inline instrumentation/opentelemetry/internal/metrics/linux_metrics.go:54]: no tests for this function? For testing this
1. either file should become a parameter to the function and then you can create custom files to test.
2. can pass an io.Reader to the function to read file. (*os.File) implements an io.Reader. During test you can use a mock version.

**Review by puneet-traceable — APPROVED** [review]

**Review by tim-mwangi — APPROVED** [review]

**Comment by thugrock7** [comment]
Tested running this by http-server/client scenario on linux, Mac, windows. Works fine
Also confirmed the exported metrics on TPA in  linux, mac os.

**Pull request merged**